dynamic_instrumentation/src/instrument_memory.rs

```diff
@@ -236,16 +236,60 @@ impl<'a, 'tcx: 'a> CollectFunctionInstrumentationPoints<'a, 'tcx> {
         find_instrumentation_def(self.tcx, self.runtime_crate_did, name)
     }
 
-    fn should_instrument(&self, local: Local, context: PlaceContext) -> bool {
-        let is_unsafe_ptr = self.body.local_decls[local].ty.is_unsafe_ptr();
-        is_unsafe_ptr && context.is_use()
+    fn is_shared_or_unsafe_ptr(&self, ty: &TyS<'tcx>) -> bool {
+        ty.is_unsafe_ptr() || (ty.is_region_ptr() && !ty.is_mutable_ptr())
```

Comment:
Should we still have a `fn should_instrument` that's updated to also return `true` if it's a shared ptr?  The intent is clearer that way.

And I think `fn is_shared_or_unsafe_ptr` could be split into `fn is_shared_ptr` and `TyS::is_unsafe_ptr`.

Reply:
i think we have too many context-dependent cases to support `should_instrument` in the form it was in

i'll add a commit to split those up

Reply:
https://github.com/immunant/c2rust/pull/521/commits/1debde1354b0bdc4c10ccb01f2d3ab16845a8469

Reply:
Okay, if they're different in enough places, then, yeah, it makes sense not to have `should_instrument`.